proposals/0569-multiline-strings.rst

```diff
@@ -0,0 +1,629 @@
+
+#. Post-process the string in the following steps:
+
+   #. Collapse line continuations
+
+   #. Remove common whitespace prefix in every line
+
+      * Ignore any characters preceding the first newline
+      * Blank lines and lines with only whitespace should not be included in this calculation
+
+   #. Remove exactly one newline from the beginning of the string (if one exists)
+
+   #. Interpret escape sequences (occurs after removing whitespace prefix so that literal ``\n`` characters are not included)
+
+#. After parsing, it becomes indistinguishable to the equivalent single-quoted string (modulo annotations for exact-printing)
+
+Line terminators will match whatever line terminators the user is using in the file.
+
+Examples
+--------
+
+Line continuations
+~~~~~~~~~~~~~~~~~~
+
+Line continuations are collapsed first and not included in the whitespace calculation
+
+::
+
+  s =
+      """
+        a b\
+    \ c d e
+        f g
+      """
+
+  -- equivalent to
+  s' = "a b c d e\nf g\n"
+
+Remove common whitespace
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the below examples, leading spaces will be marked as ``.`` for visibility.
+
+::
+
+  s =
+        """
+  ......a b c
+
+  ......d e f
+  ..
+  ....g h i
+  ......"""
+
+  -- equivalent to
+  s' = "..a b c\n\n..d e f\n\ng h i\n.."
+
+After lexing, the initial multiline above is parsed as
+
+::
+
+  [ "......a.b.c"
+  , ""
+  , "......d.e.f"
+  , ".."
+  , "....g.h.i"
+  , "......"
+  ]
+
+The blank line + the whitespace-only line are excluded from the calculation, and we calculate 4 spaces as the shared whitespace prefix, which are removed from every line.
+
+Note that the whitespace preceding the closing ``"""`` is included. This implies that there will be a trailing newline (see the "Trailing newline" example for more information).
+
+Ignore leading characters
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The common prefix calculation ignores all characters preceding the first newline. This means that characters immediately after the ``"""`` delimiter will be included verbatim. The same would occur with a line continuation (since line continuations are collapsed before the prefix calculation).
+
+::
+
+  s =
+    """Line 1
+       Line 2
+    Line 3
+    """
+
+  s_2 =
+    """\
+   \Line 1
+       Line 2
+    Line 3
+    """
+
+  -- equivalent to
+  s' = "Line 1\n   Line 2\nLine 3"
+
+This implies that normal strings could also be written using ``"""``
+
+::
+
+  -- the following are equivalent
+  s = """hello world"""
+  s' = "hello world"
+
+Mixing tabs and spaces
+~~~~~~~~~~~~~~~~~~~~~~
+
+Per the `Haskell report <https://www.haskell.org/onlinereport/haskell2010/haskellch10.html#x17-17800010.3>`_, tabs will be treated as 8 spaces in the calculation. If the common prefix is in the middle of a tab (which would only happen if mixing spaces and tabs), the tab will be converted to 8 spaces first. But don't do this.
```

Comment:
For what it's worth, it's not what the report says. It says that there is a _tab stop_ every 8 characters.

Reply:
Fair. Is there a concrete suggestion here, or just a quibble on wording?

Reply:
I would expect it to make a difference in how you compute the common prefix when there is a mixture of tabs and spaces. For example (`.` = space, `>` = tab)

```
........
.>
....>
```

I think the current wording of the proposal would treat the first line as 8 spaces, the second as 9, and the third as 12. But the wording of the Report as per @aspiwack would treat all three lines as 8 spaces.

Reply:
I agree with Eric that this could be made more precise.

Note that GHC already has a warning if tabs are used. Not on by default, but I take it as indication that it's a generally accepted fact that tabs in indentation sensitive code are a bad idea.

So I'd be open to simply disallow tabs in the common prefix stripped here, before we come up with complicated rules when tabs and spaces are mixed, and tabs have to be ”broken up” etc.

Reply:
Ah sorry, @aspiwack + @gridaphobe, I see what you're saying now. Yes, that's a good point. What about this: convert all tabs into spaces as a first step, but instead of just 1 tab = 8 spaces, fill in the number of spaces until the next tab stop.

If the user really wants to have literal tabs in the string, they can add literal `\t` characters to the string.

@nomeata I don't think we should just disallow tabs because the point of the common prefix is that devs want to keep the multiline string at the same indentation level of surrounding code, and switching to spaces would ruin the indentation compared with the tab-indented lines.

Reply:
> convert all tabs into spaces as a first step, but instead of just 1 tab = 8 spaces, fill in the number of spaces until the next tab stop

Yeah, that’s what I would expect. Two clarifications neeeded, though: what about
```
␣␣␣␣␣␣␣␣begin of indentation
␣␣␣␣␣␣␣␣⇥       this was a tab
```
i.e. tabs _after_ the common prefix. Preserved as `\t` or converted to spaces?

And what about
```
␣␣␣␣begin of indentation
␣␣␣␣⇥    this was a tab
```
The common prefix is clearly `␣␣␣␣`, but removing that would mean the tabstop (which is not part of the common prefix) no longer means the same thing – it’ll be worth 8 instead of 4 spaces after stripping the common prefix.


> @nomeata I don't think we should just disallow tabs because the point of the common prefix is that devs want to keep the multiline string at the same indentation level of surrounding code, and switching to spaces would ruin the indentation compared with the tab-indented lines.

The point would be that if you want to use this new feature, you better switch to space-only indentation for the whole file first. But maybe we are not up for surviving the shitstorm that this may cause…

Reply:
You first said “all tabs”, now “all leading tabs” – so a tab that’s not leading would remain?  Or would tabs further to the right remain?

Sorry to ask these annoying questions, but we have to nail them down of course.


I’m personally fine with “All tabs are replaced by spaces, consistently with Haskell’s indentation rules” as a first step, before `Collapse line continuations` (I assume doing it before is the right thing? May not matter much).

Reply:
Yes, it would be a very first step. And I meant only leading tabs; I'm looking at the Haskell report, and it looks like tabs are not allowed in normal strings, is that right? So the tab rules would be:
1. Leading tabs immediately converted to spaces
2. Non-leading tabs are a syntax error (or whatever the error is when using tabs in a normal string)

Reply:
Updated: https://github.com/ghc-proposals/ghc-proposals/pull/569/commits/e5eb5669ad68ceb1e1cf49a7cf65c1d33b22b5ed